File: InterviewPreparationKit/2DArray.cpp
// https://www.hackerrank.com/challenges/2d-array/problem
#include <bits/stdc++.h>

using namespace std;

// Complete the hourglassSum function below.
int hourglassSum(vector<vector<int>> arr) {
    int max = INT_MIN;

    for (int i = 0; i < 16; i++) {
        int first = arr[i / 4][i % 4];
        int second = arr[i / 4][(i % 4) + 1];
        int third = arr[i / 4][(i % 4) + 2];
        int fourth = arr[(i / 4) + 1][(i % 4) + 1];
        int fifth = arr[(i / 4) + 2][i % 4];
        int sixth = arr[(i / 4) + 2][(i % 4) + 1];
        int seventh = arr[(i / 4) + 2][(i % 4) + 2];

        int sum = first + second + third + fourth + fifth + sixth + seventh;
        if (sum > max) {
            max = sum;
        }
    }

    return max;
}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    vector<vector<int>> arr(6);
    for (int i = 0; i < 6; i++) {
        arr[i].resize(6);

        for (int j = 0; j < 6; j++) {
            cin >> arr[i][j];
        }

        cin.ignore(numeric_limits<streamsize>::max(), '\n');
    }

    int result = hourglassSum(arr);

    fout << result << "\n";

    fout.close();

    return 0;
}

File: InterviewPreparationKit/Arrays/2DArray.js
// https://www.hackerrank.com/challenges/2d-array/problem
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.replace(/\s*$/, '')
        .split('\n')
        .map(str => str.replace(/\s*$/, ''));

    main();
});

function readLine() {
    return inputString[currentLine++];
}

// Complete the hourglassSum function below.
function hourglassSum(arr) {
    let max = Number.MIN_SAFE_INTEGER;

    for (let i = 0; i < 16; i++) {
        let first = arr[parseInt(i / 4)][i % 4];
        let second = arr[parseInt(i / 4)][(i % 4) + 1];
        let third = arr[parseInt(i / 4)][(i % 4) + 2];
        let fourth = arr[parseInt(i / 4) + 1][(i % 4) + 1];
        let fifth = arr[parseInt(i / 4) + 2][i % 4];
        let sixth = arr[parseInt(i / 4) + 2][(i % 4) + 1];
        let seventh = arr[parseInt(i / 4) + 2][(i % 4) + 2];

        let sum = first + second + third + fourth + fifth + sixth + seventh;
        if (sum > max) {
            max = sum;
        }
    }

    return max;
}

function main() {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    let arr = Array(6);

    for (let i = 0; i < 6; i++) {
        arr[i] = readLine().split(' ').map(arrTemp => parseInt(arrTemp, 10));
    }

    let result = hourglassSum(arr);

    ws.write(result + "\n");

    ws.end();
}
